### PR TITLE
HIL integration for Gazebo

### PR DESCRIPTION
**Pull request by arwagoner**
This PR integrates HIL for Gazebo. Gazebo listens to a serial port for a connection with the FCU and establishes a UDP proxy to QGC.

### CODE REVIEW

**Comment by TSC21** [comment]
Looks good @arwagoner! Can you please have a try on https://github.com/PX4/Firmware/pull/8532 and check if it does work with current master? Thanks

File: include/gazebo_mavlink_interface.h [hunk 1]
```diff
@@ -17,6 +17,23 @@
...
+#include <deque>
+#include <atomic>
+#include <chrono>
+#include <thread>
+#include <memory>
+#include <sstream>
+#include <cassert>
+#include <stdexcept>
+#include <boost/asio.hpp>
+#include <boost/bind.hpp>
```
Comment (TSC21) [inline include/gazebo_mavlink_interface.h]: double include of `bind.hpp` - look at line 38.

File: models/rotors_description/urdf/iris_base.xacro [hunk 2]
```diff
@@ -7,6 +7,13 @@
   <xacro:arg name='name' default='iris' />
   <xacro:arg name='mavlink_addr' default='INADDR_ANY' />
   <xacro:arg name='mavlink_udp_port' default='14560' />
+  <xacro:arg name='serial_enabled' default='false' />
```
Comment (TSC21) [inline models/rotors_description/urdf/iris_base.xacro:10]: we are missing these parameters on the SDF of the other files.

File: include/gazebo_mavlink_interface.h [hunk 3]
```diff
@@ -105,7 +141,19 @@ class GazeboMavlinkInterface : public ModelPlugin {
...
+        tx_q {},
+        rx_buf {},
+        m_status {},
+        m_buffer {},
+        io_service(),
+        serial_dev(io_service),
+        device_(kDefaultDevice),
+        baudrate_(kDefaultBaudRate),
+        hil_mode_(false),
+        hil_state_level_(true)
```
Comment (TSC21) [inline include/gazebo_mavlink_interface.h]: `hil_state_level_(false)`?

Comment (arwagoner) [reply include/gazebo_mavlink_interface.h]: Yes. I agree.

**Comment by LorenzMeier** [comment]
Nice!

**Comment by hamishwillee** [comment]
Can we get some docs on exact idiot (me) proof setup? 

Thoughts ....
- I think we'd update [Hardware in the Loop Simulation (HITL)](https://dev.px4.io/en/simulation/hitl.html) with a section. However if there is a lot of configuration then maybe a "linking" dedicated document.
- Should list what airframes are supported for use with HIL
- Is there a specific Gazebo version this works with.
- Is there anything special required to also get this working with ROS?
- Why would you use this, say, instead of using Xplane or similar?

**Comment by TSC21** [comment]
> - I think we'd update [Hardware in the Loop Simulation (HITL)](https://dev.px4.io/en/simulation/hitl.html) with a section. However if there is a lot of configuration then maybe a "linking" dedicated document.

Agree.

- Should list what airframes are supported for use with HIL

We can test them all but I believe that this can work with all the airframes that have a correspondent model to use with SITL Gazebo

> - Is there a specific Gazebo version this works with.

Currently working with Gazebo 7. We have not tried with above versions.

> - Is there anything special required to also get this working with ROS?

What type of interface are you looking for in ROS? I believe it should be possible to setup a UDP bind point in the Gazebo side where we can send, for example, setpoint commands and retrieve that from the FCU, as it's done with QGC. I also have pending the PR for Gazebo interface with MAVROS, which would use MAVROS as the data proxy/interface, but me and @arwagoner bet on the serial interface approach for Gazebo, so that feature turns out to be secondary atm.

> - Why would you use this, say, instead of using Xplane or similar?

Easy: This allows to setup vehicle models more easily (we still to doc this of course) and test a more extensive list of airframes.

**Comment by hamishwillee** [comment]
Thanks!

>> Is there a specific Gazebo version this works with.

> Currently working with Gazebo 7. We have not tried with above versions.

OK, so Gazebo 8 is the "supported version" with PX4 (and in our [installation instructions](https://dev.px4.io/en/simulation/gazebo.html#installation) and standard packaging) so it would be good to test with that too. 

We do use Gazebo 7 for setup with [ROS](https://dev.px4.io/en/simulation/ros_interface.html), because that is the default compatible version that is included in ROS Kinetic instructions.

>> Is there anything special required to also get this working with ROS?

> What type of interface are you looking for in ROS?

My job is technical documentation. I'm not so concerned in the first instance whether ROS is supported or not - I just want to be able to make clear statements about what can an cannot be done. Going forward though as a developer advocate I want us to  have a great experience for ROS users - but I don't have the experience to say what that is :-)

@TSC21 I presume you are acting as advocate for ROS users in this discussion?

**Comment by TSC21** [comment]
> @TSC21 I presume you are acting as advocate for ROS users in this discussion?

Yeah I can fit that position if required :)

**Comment by hamishwillee** [comment]
>> @TSC21 I presume you are acting as advocate for ROS users in this discussion?

> Yeah I can fit that position if required :)

Good, and then you can help me tidy the associated docs :-) No rest for the wicked!

**Comment by arwagoner** [comment]
@TSC21, I just confirmed that this does work with the current master after the changes made in https://github.com/PX4/Firmware/pull/8532. 
gazebo_hitl

**Comment by TSC21** [comment]
> @TSC21, I just confirmed that this does work with the current master after the changes made in https://github.com/PX4/Firmware/pull/8532. 
gazebo_hitl

Awesome news!

**Comment by TSC21** [comment]
@arwagoner we are in a process of getting CI working for SITL Gazebo, so let's just get CI working and then we can merge this.

**Comment by TSC21** [comment]
@arwagoner can you please rebase on master?

**Comment by LorenzMeier** [comment]
The diff of this PR looks bogus - I think something is wrong with the GIT history. I've squashed it in this branch: https://github.com/PX4/sitl_gazebo/pull/157

**Comment by TSC21** [comment]
Replaced in #157

**Pull request closed without merging**